python/src/iceberg/partition_spec.py

```diff
@@ -0,0 +1,289 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from urllib import parse
+from collections import defaultdict
+from typing import List, Optional
```

Comment:
QQ: are defaultdict and List needed?

Reply:
Using `dict` and `list` as generics is only possible in python 3.9+ (added in [PEP 585](https://www.python.org/dev/peps/pep-0585/)).For that to work, we would have to add `from __future__ import annotations` which doesn't look better imo.

Although I am curious why `defaultdict` is used here instead of a standard `dict` and using `Dict` from typing for the typehints. From what I can see, the default lambda used (which is `list`) is never called in this scenario I see in `_generate_fields_by_source_id()`:
```py
            fields_source_to_field_dict = defaultdict(list)
            for field in self.fields:
                fields_source_to_field_dict[field.source_id] = [field]
```